s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java

```diff
@@ -0,0 +1,249 @@
+      checkArgument(!isNullOrEmpty(hostname), "hostname should not be null or empty.");
+      return new S2AProtocolNegotiationHandler(
+        grpcHandler, channelPool, localIdentity, hostname, service);
+    }
+
+    @Override
+    public void close() {
+      service.shutdown();
+      channelPool.close();
+    }
+  }
+
+  @VisibleForTesting
+  static class BufferReadsHandler extends ChannelInboundHandlerAdapter {
+    private final List<Object> reads = new ArrayList<>();
+    private boolean readComplete;
+
+    public List<Object> getReads() {
+      return reads;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext unused, Object msg) {
+      reads.add(msg);
+    }
+
+    @Override
+    public void channelReadComplete(ChannelHandlerContext unused) {
+      readComplete = true;
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+      for (Object msg : reads) {
+        super.channelRead(ctx, msg);
+      }
+      if (readComplete) {
+        super.channelReadComplete(ctx);
+      }
+    }
+  }
+
+  private static final class S2AProtocolNegotiationHandler extends ProtocolNegotiationHandler {
+    private final S2AChannelPool channelPool;
+    private final Optional<S2AIdentity> localIdentity;
+    private final String hostname;
+    private final GrpcHttp2ConnectionHandler grpcHandler;
+    private final ListeningExecutorService service;
+
+    private S2AProtocolNegotiationHandler(
+        GrpcHttp2ConnectionHandler grpcHandler,
+        S2AChannelPool channelPool,
+        Optional<S2AIdentity> localIdentity,
+        String hostname,
+        ListeningExecutorService service) {
+      super(
+          // superclass (InternalProtocolNegotiators.ProtocolNegotiationHandler) expects 'next'
+          // handler but we don't have a next handler _yet_. So we "disable" superclass's behavior
+          // here and then manually add 'next' when we call fireProtocolNegotiationEvent()
+          new ChannelHandlerAdapter() {
+            @Override
+            public void handlerAdded(ChannelHandlerContext ctx) {
+              ctx.pipeline().remove(this);
+            }
+          },
+          grpcHandler.getNegotiationLogger());
+      this.grpcHandler = grpcHandler;
+      this.channelPool = channelPool;
+      this.localIdentity = localIdentity;
+      this.hostname = hostname;
+      checkNotNull(service, "service should not be null.");
+      this.service = service;
+    }
+
+    @Override
+    protected void handlerAdded0(ChannelHandlerContext ctx) {
+      // Buffer all reads until the TLS Handler is added.
+      BufferReadsHandler bufferReads = new BufferReadsHandler();
+      ctx.pipeline().addBefore(ctx.name(), /* name= */ null, bufferReads);
+
+      Channel ch = channelPool.getChannel();
+      S2AServiceGrpc.S2AServiceStub stub = S2AServiceGrpc.newStub(ch);
+      S2AStub s2aStub = S2AStub.newInstance(stub);
```

Comment:
This Stub is never closed. This causes IntegrationTest (after Larry's reordering of awaitTermination) to not complete tests quickly, as the RPC is outstanding so awaitTermination() hangs. I think once we shut down the top-level channel that releases the S2AHandshakerServiceChannel.ChannelResource after 1 second, which then calls shutdownNow() on the s2a channel which cancels the RPC.

Reply:
Yes, we never call close on the stub directly. When the 
 [close](https://github.com/grpc/grpc-java/blob/master/s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java#L144) gets invoked, everything you mentioned happens.

Reply:
That's not okay. That means we leak an RPC per connection for the life of _all_ s2a channels, which is probably the lifetime of the process.

Reply:
Would closing the stub in [close](https://github.com/grpc/grpc-java/blob/master/s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java#L146) resolve the issue?

Reply:
The channel is still needed after the SslContext future is complete (we make RPCs to S2A to verify the peer, and perform private key operations during the handshake). The `SslContextBuilder` [configures the trustManager and PrivateKeyMethod with the stub](https://github.com/grpc/grpc-java/blob/master/s2a/src/main/java/io/grpc/s2a/internal/handshaker/SslContextFactory.java#L85-L88). In this example, the channel gets returned to the pool before the handshake is complete, which causes an error when S2ATrustManager uses the stub to verify the peer cert.

The stub is no longer needed when the handshake is complete. A probably not great option would be to pass around the channel instead, and create a new stub for every RPC? Or perhaps we plumb the stub down to the ClientTlsProtocolNegotiator and close it when the [handshake is done](https://github.com/grpc/grpc-java/blob/927d21541dbc34be245f09c8180358a94ae5c031/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L644)?

Reply:
@ejona86, WDYT?

Reply:
I see. A new stub for every RPC is fine. I thought y'all _needed_ to use the same streaming RPC to preserve context. If separate RPCs are fine, that is definitely superior.

We don't want to inject into the handshake completion code itself, but that code is triggering a ProtocolNegotiationEvent. We can use that.

```java
private static final class S2AStubCleanupNegotiationHandler extends ProtocolNegotiationHandler {
  private final s2aStub; // pass in constructor
  @Override
  protected void protocolNegotiationEventTriggered(ChannelHandlerContext ctx) {
    s2aStub.close();
    fireProtocolNegotiationEvent(ctx);
    ctx.pipeline().remove(ctx);
  }
}
```

And then in S2aProtocolNegotiationHandler, `ctx.pipeline().addAfter()` that `S2AStubCleanupNegotiationHandler` before adding the tls handler. (So the TLS handler will be between that cleanup handler and the s2a handler.)

We could do alternatives like adding the tls handler _before_ the s2a handler, but they take a bit more reorganization.

Reply:
Thanks Eric! I've implemented what you mention in https://github.com/grpc/grpc-java/pull/11600

I initially did think that we could do a new stub for each RPC, since we attach the localIdentity to each RPC, but chatted with @matthewstevenson88 and he mentioned there is a requirement that the lifetime of the bidirectional stream from the client to the S2A must be tied to the lifetime of the handshake.